### PR TITLE
[SPARK-15889][SQL][STREAMING] Add a unique id to ContinuousQuery

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

ContinuousQueries have names that are unique across all the active ones. However, when queries are rapidly restarted with same name, it causes races conditions with the listener. A listener event from a stopped query can arrive after the query has been restarted, leading to complexities in monitoring infrastructure.

Along with this change, I have also consolidated all the messy code paths to start queries with different sinks.
## How was this patch tested?

Added unit tests, and existing unit tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/ContinuousQueryManager.scala [hunk 1]
```diff
@@ -170,18 +172,50 @@ class ContinuousQueryManager private[sql] (sparkSession: SparkSession) {
 
   /** Start a query */
   private[sql] def startQuery(
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/ContinuousQueryManager.scala]: Add scala docs

**Comment by tdas** [comment]
@zsxwing Can you take a look.

**Comment by zsxwing** [comment]
Could you also add `def get(id: Long): ContinuousQuery` to `ContinuousQueryManager`, and make `def get(name: String): ContinuousQuery` returns `null` for inactive queries since you are touching this file? Otherwise, LGTM.

**Comment by zsxwing** [comment]
Offline discussion: replace `get(name)` with `get(id)`.

**Comment by zsxwing** [comment]
LGTM pending tests.

**Comment by zsxwing** [comment]
Thanks. Merging to master and 2.0.

**Pull request closed without merging**